src/Storage/Services/SigningService.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            (Instance instance, long instanceInternalId) = await _instanceRepository.GetOne(instanceGuid, false, cancellationToken);
+            
+            if (instance == null) 
+            {
+                return (false, new ServiceError(404, "Instance not found"));
+            }
+
+            Application app = await _applicationRepository.FindOne(instance.AppId, instance.Org);
+
+            (bool validDataType, ServiceError serviceError) = await _applicationService.ValidateDataTypeForApp(instance.Org, instance.AppId, signRequest.SignatureDocumentDataType, instance.Process.CurrentTask?.ElementId);
+            if (!validDataType)
+            {
+                return (false, serviceError);
+            }
+
+            SignDocument signDocument = CreateSignDocument(instanceGuid, signRequest);
+
+            foreach (SignRequest.DataElementSignature dataElementSignature in signRequest.DataElementSignatures)
+            {
+                (string base64Sha256Hash, serviceError) = await _dataService.GenerateSha256Hash(instance.Org, instanceGuid, Guid.Parse(dataElementSignature.DataElementId), app.StorageAccountNumber);
+                if (string.IsNullOrEmpty(base64Sha256Hash))
+                {
+                    return (false, serviceError);
+                }
+
+                signDocument.DataElementSignatures.Add(new SignDocument.DataElementSignature
+                {
+                    DataElementId = dataElementSignature.DataElementId,
+                    Sha256Hash = base64Sha256Hash,
+                    Signed = dataElementSignature.Signed
+                });
+            }
+
+            DataElement dataElement = DataElementHelper.CreateDataElement(
+                signRequest.SignatureDocumentDataType, 
+                null, 
+                instance, 
+                signDocument.SignedTime, 
+                "application/json", 
+                $"{signRequest.SignatureDocumentDataType}.json", 
+                0, 
+                performedBy,
+                signRequest.GeneratedFromTask);
+
+            signDocument.Id = dataElement.Id;
+
+            await DeleteExistingSignDocument(instance, signRequest.SignatureDocumentDataType, signDocument); 
+        
+            using (var fileStream = new MemoryStream(System.Text.Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(signDocument, Formatting.Indented))))
+            {
+                await _dataService.UploadDataAndCreateDataElement(instance.Org, fileStream, dataElement, instanceInternalId, app.StorageAccountNumber);
+            }
+            
+            await _instanceEventService.DispatchEvent(InstanceEventType.Signed, instance);
+            return (true, null);
+        }
+
+        private async Task DeleteExistingSignDocument(Instance instance, string signDocDataType, SignDocument newSignDocument)
```

Comment:
Last parameter could be of type Signee. Maybe change method name as well:
```suggestion
        private async Task DeleteExistingSignDocumentForSignee(Instance instance, string signDocDataType, Signee signee)
```